### PR TITLE
openjdk: 14.0.2 -> 15.0.1

### LINKED ISSUE #98784: Request: OpenJDK 15
I am missing version 15 of the `openjdk` package, see https://jdk.java.net/15/

### PR DESCRIPTION
**Pull request by jerith666**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I am off work until early January, so won't be able to confirm this survives a few workdays until then.  :)

continues from #83116.  fixes #98784.

cc @edwtjo @shlevy @NeQuissimus @lopsided98 @xfix @petabyteboy @lopsided98 @bachp

### CODE REVIEW

File: pkgs/development/compilers/openjdk/14.nix [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  openjdk = stdenv.mkDerivation rec {
+    pname = "openjdk" + lib.optionalString headless "-headless";
+    version = "${major}${update}${build}";
+
+    src = fetchurl {
+      url = "http://hg.openjdk.java.net/jdk-updates/jdk${major}u/archive/jdk-${version}.tar.gz";
+      sha256 = "1s1pc6ihzf0awp4hbaqfxmbica0hnrg8nr7s0yd2hfn7nan8xmf3";
+    };
+
+    nativeBuildInputs = [ pkgconfig autoconf ];
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/openjdk/14.nix:25]:
```suggestion
    nativeBuildInputs = [ pkg-config autoconf ];
```

File: pkgs/development/compilers/openjdk/14.nix [hunk 2]
```diff
@@ -0,0 +1,157 @@
+{ stdenv, lib, fetchurl, bash, pkgconfig, autoconf, cpio, file, which, unzip
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/openjdk/14.nix:1]:
```suggestion
{ stdenv, lib, fetchurl, bash, pkg-config, autoconf, cpio, file, which, unzip
```

File: pkgs/development/compilers/openjdk/14.nix [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  major = "14";
+  update = ".0.2";
+  build = "-ga";
+
+  openjdk = stdenv.mkDerivation rec {
+    pname = "openjdk" + lib.optionalString headless "-headless";
+    version = "${major}${update}${build}";
+
+    src = fetchurl {
+      url = "http://hg.openjdk.java.net/jdk-updates/jdk${major}u/archive/jdk-${version}.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/openjdk/14.nix:21]:
```suggestion
      url = "https://hg.openjdk.java.net/jdk-updates/jdk${major}u/archive/jdk-${version}.tar.gz";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Ofborg fails because jdk14 is used as an input here https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/apache-directory-studio/default.nix#L1.

File: pkgs/development/compilers/openjdk/14.nix (hunk 1 above)
Comment (jerith666) [reply pkgs/development/compilers/openjdk/14.nix]: fixed

File: pkgs/development/compilers/openjdk/14.nix (hunk 2 above)
Comment (jerith666) [reply pkgs/development/compilers/openjdk/14.nix]: fixed

File: pkgs/development/compilers/openjdk/14.nix (hunk 3 above)
Comment (jerith666) [reply pkgs/development/compilers/openjdk/14.nix]: fixed

**Comment by jerith666** [comment]
`mindustry` also uses `jdk14` explicitly.  Confirmed that it and `apache-directory-studio` launch cleanly.

**Comment by doronbehar** [comment]
@jerith666 There are 282 packages that will get affected by this:

```
curl --silent https://gist.githubusercontent.com/GrahamcOfBorg/01e59afc3364e0b5a5add261f9a92a52/raw/e48220f36edf6cfa96080ddd954debe17228fc74/Changed%2520Paths | awk '{print $2}' | sort -u | wc -l
```

Unfortunately, this kind of major version updates require a lot of manual testings, especially for jdk, see https://github.com/NixOS/nixpkgs/pull/89731 . Ideally, we should test each package individually and make sure it launches in runtime, but it's probably a lot of work.. Testing 10% percent of these packages will be persuasive and should help this PR get accepted.

**Comment by petabyteboy** [comment]
### nix-review output

11 packages marked as broken and skipped:
aldor foo-yc20 jboss junit junixsocket octave-jit openfire pig plm purePackages.faust spark

2 packages failed to build:
graalvm8 jvmci8

251 packages built:
DisnixWebService alchemy alda ammonite ammonite_2_12 apache-directory-studio apache-flex-sdk apktool arduino arduino-core asciidoc-full-with-plugins asciidoctorj aspectj astrolabe-generator atlassian-cli avro-tools aws_mturk_clt awstats axis2 basex bfg-repo-cleaner bftools bitwig-studio bitwig-studio1 bitwig-studio2 bloop boot briss ccemux cfr checkstyle clojure-lsp clooj closurecompiler confluent-platform coursier davmail dbeaver dex2jar ditaa ec2_api_tools emem emscripten epubcheck ergo exhibitor faust faust2alqt faust2alsa faust2csound faust2firefox faust2jack faust2jackrust faust2jaqt faust2ladspa faust2lv2 faustlive flexibee flink flink_1_5 flyway frostwire frostwire-bin gama ganttproject-bin geoipjava gitAndTools.subgit gogui gpsprune graylog gremlin-console groove groovy h2 haskellPackages.hzk haxePackages.hxjava hsqldb html5validator hydraAntLogger imagej ino ipscan jadx jameica jbake jbidwatcher jd-gui jdiskreport jdk jre_headless jedit jflex jicofo jing jitsi-videobridge jmeter jnetmap josm jre_minimal jruby jugglinglab jython kapitonov-plugins-pack key keycloak kodi kodi-gbm kodi-wayland kodiGBM kodiPlain kodiPlainWayland kodiPlugins.inputstream-adaptive kodiPlugins.joystick kodiPlugins.kodi-platform kodiPlugins.pvr-hdhomerun kodiPlugins.pvr-hts kodiPlugins.pvr-iptvsimple kodiPlugins.steam-controller kodiPlugins.vfs-libarchive kodiPlugins.vfs-sftp kotlin ktlint languagetool leiningen leo3-bin liquibase logisim logstash logstash6-oss logstash7 logstash7-oss lumo ma1sd macse makemkv martyr marvin mediathekview metabase metals micronaut mill minc_widgets mindustry mindustry-server minecraft minecraft-server mkgmap mkgmap-splitter mooSpace msgviewer munin mxisd nailgun netlogo nexus octave octaveFull omegat opa open-ecard openapi-generator-cli openapi-generator-cli-unstable openems opengrok openjdk15-bootstrap openjfx15 openjump papermc perl530Packages.NetZooKeeper perl532Packages.NetZooKeeper picard-tools picolisp plantuml plantuml-server portmod prometheus-jmx-httpserver python37Packages.databricks-connect python37Packages.python-csxcad python37Packages.python-openems python37Packages.sphinxcontrib_plantuml python38Packages.databricks-connect python38Packages.python-csxcad python38Packages.python-openems python38Packages.sphinxcontrib_plantuml python39Packages.databricks-connect python39Packages.python-csxcad python39Packages.python-openems python39Packages.sphinxcontrib_plantuml quantomatic rascal riemann runelite saxon saxon-he saxonb saxonb_9_1 sbt sbt-extras sbt-with-scala-native scalafmt schemaspy selendroid selenium-server-standalone seyren shattered-pixel-dungeon signal-cli smc snpeff soapui solr sonar-scanner-cli spring spring-boot-cli springLobby squirrel-sql sshoogr stm32cubemx storm subsonic swagger-codegen swingsane swt tabula tambura teensyduino tessera tomcat-native tomcat_connectors tr-patcher tuxguitar tvbrowser-bin umlet varscan vassal vimPlugins.lens-vim vimPlugins.vim-grammarous visualvm vue vuze weka workcraft xmind xmloscopy yed youtrack yuicompressor zipkin zkfuse zookeeper_mt zxing

graalvm8 and jvmci8 fail with `Could not find a JDK >=1.8.0_141 and <1.9` and this is also the case on master, so no new breakage there.

---

### manual testing

Packages that I tested and that worked fine:

- jd-gui
- frostwire
- minecraft
- minecraft-server
- mindustry
- mindustry-server
- shattered-pixel-dungeon
- mxisd (via nixos/tests/mxisd.nix)
- netlogo
- plantuml
- spring (kinda, I got into the menu but could not get ingame, same happens on master so it should be fine)
- selenium-server
- umlet

**Packages that didn't work and work on master:**

- arduino: Crashes after splash screen, works with jdk14
- jitsi-videobridge (via nixos/tests/jitsi-meet.nix): Unrecognized VM option 'UseConcMarkSweepGC', see https://github.com/jitsi/jitsi-videobridge/issues/1547

Packages that didn't work but also didn't work with master either:

- tuxguitar: SIGSEGV in jre
- vuze: java.lang.UnsupportedClassVersionError (it's being compiled with jdk14/jdk15 but run with jre/jdk8)
- tvbrowser-bin: java.lang.ClassNotFoundException

---

### summary

The only packages I found that didn't work here but did work on master were arduino and jitsi-videobridge. There seems to be an update available for arduino so maybe that will fix it. I'm taking care of the GC flags in jitsi videobridge and opened a bug report upstream at https://github.com/jitsi/jitsi-videobridge/issues/1547.

A lot of the affected packages are duplicates, as in multiple derivations used for the same application, for example Kodi and its plugins make up 14 derivations.

File: pkgs/development/compilers/openjdk/openjfx/15.nix [hunk 4]
```diff
@@ -64,8 +64,10 @@ let
     outputHashMode = "recursive";
     # Downloaded AWT jars differ by platform.
     outputHash = {
-      x86_64-linux = "077zss95iq6iskx7ghz1c57ymydpzj0wm7r1pkznw99l9xwvdmqi";
-      i686-linux = "03gglr2sh77cyg16qw9g45ji33dg7i93s5s30hz3mh420g112qa0";
+      x86_64-linux = "0hmyr5nnjgwyw3fcwqf0crqg9lny27jfirycg3xmkzbcrwqd6qkw";
+      # The build-time dependencies don't currently build for i686, so no
+      # reason to fetch this one correctly either...
+      i686-linux = "0000000000000000000000000000000000000000000000000000";
```
Comment (doronbehar) [inline pkgs/development/compilers/openjdk/openjfx/15.nix:70]: Hash:

```suggestion
      i686-linux = "0000000000000000000000000000000000000000000000000000";
```

**Comment by doronbehar** [comment]
> **Packages that didn't work and work on master:**
> 
>     * arduino: Crashes after splash screen, works with jdk14
> 
>     * jitsi-videobridge (via nixos/tests/jitsi-meet.nix): Unrecognized VM option 'UseConcMarkSweepGC', see [jitsi/jitsi-videobridge#1547](https://github.com/jitsi/jitsi-videobridge/issues/1547)

Thanks for the report. So we should add commits that will pin them to jdk14.

File: pkgs/development/compilers/openjdk/openjfx/15.nix (hunk 4 above)
Comment (jerith666) [reply pkgs/development/compilers/openjdk/openjfx/15.nix:70]: These lines are identical according to the GitHub UI ... what is the change you're suggesting here?

Comment (doronbehar) [reply pkgs/development/compilers/openjdk/openjfx/15.nix:70]: Never mind :sweat_smile: . Haven't read the comment.

**Comment by doronbehar** [comment]
@jerith666 if you could please take care of `arduino` and `jitsi-videobridge`.

**Comment by jerith666** [comment]
I just tried `arduino` and it works okay for me locally:

```
$ git rev-parse HEAD
d9d16e27c8ed6780c8486094ef95e8c0a38fe3ec

$ ls -l arduino-jdk1*
lrwxrwxrwx 1 matt users 58 Dec 25 15:44 arduino-jdk14 -> /nix/store/4hk2g10wxvw641gyvzp2l4pw304s18z8-arduino-1.8.12
lrwxrwxrwx 1 matt users 58 Dec 25 16:11 arduino-jdk15 -> /nix/store/8ajpqj2yaypg7khs3fa08rxlf4pmyckn-arduino-1.8.12
```

I get past the splash screen and was able to load and compile an example in both of those builds.  Here's the output from the one using jdk 15:

```
$ ./arduino-jdk15/bin/arduino 
Picked up JAVA_TOOL_OPTIONS: 
Set log4j store directory /home/matt/.arduino15

(java:5114): Gtk-WARNING **: 16:16:52.623: Theme parsing error: colors.css:71:44: Invalid number for color value

(java:5114): Gtk-WARNING **: 16:16:52.623: Theme parsing error: colors.css:72:44: Invalid number for color value

(java:5114): Gtk-WARNING **: 16:16:52.623: Theme parsing error: colors.css:74:53: Invalid number for color value

(java:5114): Gtk-WARNING **: 16:16:52.623: Theme parsing error: colors.css:75:53: Invalid number for color value

(java:5114): Gtk-WARNING **: 16:16:52.623: Theme parsing error: colors.css:76:56: Invalid number for color value

(java:5114): Gtk-WARNING **: 16:16:52.623: Theme parsing error: colors.css:77:65: Invalid number for color value
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by processing.app.linux.GTKLookAndFeelFixer (file:/nix/store/8ajpqj2yaypg7khs3fa08rxlf4pmyckn-arduino-1.8.12/share/arduino/lib/arduino-core.jar) to field com.sun.java.swing.plaf.gtk.GTKLookAndFeel.styleFactory
WARNING: Please consider reporting this to the maintainers of processing.app.linux.GTKLookAndFeelFixer
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

Sketch uses 1860 bytes (5%) of program storage space. Maximum is 32256 bytes.
Global variables use 188 bytes (9%) of dynamic memory, leaving 1860 bytes for local variables. Maximum is 2048 bytes.
```

@petabyteboy said "I'm taking care of the GC flags in jitsi videobridge", so I'll wait for that but I'm sure I can figure it out myself if needed.  :)

**Comment by jerith666** [comment]
Just force-pushed a new commit that includes a patch for `jitsi-videobridge`.  `nixos/tests/jitsi-meet.nix` passes now.

**Review by mmilata — APPROVED** [review]
Didn't test but Jitsi changes look reasonable.

**Review by doronbehar — APPROVED** [review]
I too didn't test a thing but all changes look good and I believe the OP has tested what they tested to be functional.

**Comment by jerith666** [comment]
I've got two workdays under my belt with this now (running Eclipse on top of this JDK), and no issues to report.  So, removing the WIP label; I think this is ready to merge.

**Comment by doronbehar** [comment]
OK. We can't wait for such major updates for ever. Let's cook this in unstable :).

**Pull request merged**